File: partiql-plan/src/main/java/org/partiql/plan/rel/RelOffset.java
package org.partiql.plan.rel;

import org.jetbrains.annotations.NotNull;
import org.partiql.plan.Operand;
import org.partiql.plan.OperatorVisitor;
import org.partiql.plan.rex.Rex;

import java.util.List;

/**
 * Logical offset abstract base class.
 */
public abstract class RelOffset extends RelBase {

    /**
     * @return new {@link RelOffset} instance
     */
    @NotNull
    public static RelOffset create(@NotNull Rel input, @NotNull Rex offset) {
        return new Impl(input, offset);
    }

    /**
     * @return input rel (operand 0)
     */
    @NotNull
    public abstract Rel getInput();

    /**
     * @return offset rex (operand 1)
     */
    @NotNull
    public abstract Rex getOffset();

    @NotNull
    @Override
    protected final RelType type() {
        return getInput().getType();
    }

    @NotNull
    @Override
    protected final List<Operand> operands() {
        Operand c0 = Operand.single(getInput());
        Operand c1 = Operand.single(getOffset());
        return List.of(c0, c1);
    }

    @Override
    public <R, C> R accept(@NotNull OperatorVisitor<R, C> visitor, C ctx) {
        return visitor.visitOffset(this, ctx);
    }

    @NotNull
    public abstract RelOffset copy(@NotNull Rel input);

    @NotNull
    public abstract RelOffset copy(@NotNull Rel input, @NotNull Rex offset);

    private static class Impl extends RelOffset {

        private final Rel input;
        private final Rex offset;

        private Impl(Rel input, Rex offset) {
            this.input = input;
            this.offset = offset;
        }

        @NotNull
        @Override
        public Rel getInput() {
            return input;
        }

        @NotNull
        @Override
        public Rex getOffset() {
            return offset;
        }

        @NotNull
        @Override
        public RelOffset copy(@NotNull Rel input) {
            return new Impl(input, offset);
        }

        @NotNull
        @Override
        public RelOffset copy(@NotNull Rel input, @NotNull Rex offset) {
            return new Impl(input, offset);
        }
    }
}

File: partiql-plan/src/main/kotlin/org/partiql/plan/rel/RelOffset.kt
package org.partiql.plan.rel

import org.partiql.plan.rex.Rex

/**
 * Logical `OFFSET` operator.
 */
public interface RelOffset : Rel {

    public fun getInput(): Rel

    public fun getOffset(): Rex

    override fun getChildren(): Collection<Rel> = listOf(getInput())

    override fun getSchema(): org.partiql.plan.Schema = getInput().getSchema()

    override fun isOrdered(): Boolean = getInput().isOrdered()

    override fun <R, C> accept(visitor: RelVisitor<R, C>, ctx: C): R = visitor.visitOffset(this, ctx)
}

/**
 * Default [RelOffset] implementation.
 */
internal class RelOffsetImpl(input: Rel, offset: Rex) : RelOffset {

    // DO NOT USE FINAL
    private var _input: Rel = input
    private var _offset: Rex = offset

    override fun getInput(): Rel = _input

    override fun getOffset(): Rex = _offset

    override fun getChildren(): Collection<Rel> = listOf(_input)

    override fun getSchema(): org.partiql.plan.Schema = _input.getSchema()

    override fun isOrdered(): Boolean = _input.isOrdered()

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is RelOffset) return false
        if (_input != other.getInput()) return false
        if (_offset != other.getOffset()) return false
        return true
    }

    override fun hashCode(): Int {
        var result = 1
        result = 31 * result + _input.hashCode()
        result = 31 * result + _offset.hashCode()
        return result
    }
}
